ipa-core/src/net/test.rs

```diff
@@ -140,31 +251,75 @@ impl TestConfigBuilder {
         self
     }
 
-    #[must_use]
-    pub fn build(self) -> TestConfig {
-        let mut sockets = None;
-        let ports = self.ports.unwrap_or_else(|| {
-            let socks = array::from_fn(|_| TcpListener::bind("localhost:0").unwrap());
-            let ports = socks
-                .each_ref()
-                .map(|sock| sock.local_addr().unwrap().port());
-            sockets = Some(socks);
-            ports
+    /// Creates 3 MPC test servers, assigning ports from OS if necessary.
+    fn build_mpc_servers(
+        &self,
+        optional_ports: Option<Vec<u16>>,
+        ring_index: usize,
+    ) -> TestServers {
+        let mut sockets = vec![None, None, None];
+        let ports = optional_ports.unwrap_or_else(|| {
+            sockets = (0..3)
+                .map(|_| Some(TcpListener::bind("localhost:0").unwrap()))
+                .collect();
+            sockets
+                .iter()
+                .map(|sock| sock.as_ref().unwrap().local_addr().unwrap().port())
+                .collect()
         });
-        let (scheme, certs) = if self.disable_https {
-            ("http", [None, None, None])
+        // Ports will always be defined after this. Socks only if there were no ports set.
```

Comment:
Not sure to what part you're referring...

But this function creates 3 servers.

Reply:
After line 270 all the port situation will be set. Either you gave it as param or we asked the os.

Not sure to what part you're referring when you say "Why we use all ports available to build servers"...  But this function creates 3 servers.

Reply:
The confusion comes from either using `optional_ports` or synthesizing `ports` inside 

`optional_ports: Option<Vec<u16>>` which can be arbitrary size according to this function (it does not enforce size to be 3, if it should, then it would be better to take an array). So `ports` may have variable size